WordPress/gutenberg
# Reset focus after route navigation in Site Editor

## Description

Part of https://github.com/WordPress/gutenberg/issues/36597. Alternative to https://github.com/WordPress/gutenberg/pull/37265.

Continue from #36488, this PR refactors a large part of the original PR to optimize the render times needed when switching to another route.

This PR will reset the focus to the navigation toggle button on route change. In order to prevent it from focusing on the button on initial page load, we create a flag `isInitialPageLoadRef` and set it to false after the first render.

https://github.com/WordPress/gutenberg/blob/bdb289a158a6f4dd41c20bcc1a8f5eeec5e5963c/packages/edit-site/src/components/routes/use-reset-focus-on-route-change.js#L13-L20

However, the button will remount whenever we switch from list page to the editor or the other way around. It causes `isInitialPageLoadRef` to be reset to `true` after remount, thus skipping the focus.

We could do some weird hack to make it work with minimal efforts, but a more robust fix will be to prevent the button from remounting in the first place. The button only remounts because we change the "parent" of it after switching to a different route. For instance, switching 
from `<List><NavigationToggle /></List>` to `<Editor><NavigationToggle /></Editor>`.

It's called ["Reparenting"](https://github.com/facebook/react/issues/3965) in React, changing the parent of an element will inevitably remount itself. There are only a few options we could do to bypass that. The basic ideas are the same though, **don't re-parent the element**.

The approach this PR takes is to use a `renderLayout` function to render the route layout rather than using yet another component. The advantage of render function here is that it doesn't count as a _parent_, since there's no component being rendered. The disadvantage is that we can't use hooks inside the render function, so we'll have to pass the state from its parent.

First, we define the `renderLayout` function on the Route entry component:

https://github.com/WordPress/gutenberg/blob/bdb289a158a6f4dd41c20bcc1a8f5eeec5e5963c/packages/edit-site/src/components/list/index.js#L20-L55

Then, we just have to call the render function with all the required states in the parent (`<App>`):

https://github.com/WordPress/gutenberg/blob/bdb289a158a6f4dd41c20bcc1a8f5eeec5e5963c/packages/edit-site/src/components/app/index.js#L46-L49

Since that `<Layout>` will never re-parent, neither will `<InterfaceSkeleton>` and `<NavigationToggle>` underneath it, they will not remount on route change anymore. The fix for focus reset then become trivial and predictable.

Note that this refactor doesn't just benefit this focus fix, but also improves the overall performance of the editor, since that it doesn't have to do unnecessary remounts anymore.

This is a big change though, so if we can't get enough testing, I agree to postpone it after 5.9. We can fix focus management in another smaller PR (like #37265).

## How has this been tested?



1. Activate `tt1-blocks` theme or any other block-based theme
2. Visit Appearance -> Editor
3. On initial page load, the focus shouldn't focus on the navigation toggle button (should it?)
4. Toggle the navigation button to open the sidebar
5. Visit the "Templates" or "Template Parts" link
6. After the navigation, the focus should reset to the navigation toggle button
7. Select any item in the list page to go to the editor
8. After the navigation, the focus should reset to the navigation toggle button

## Types of changes




Bug fix

## Checklist:
